tests/garage/tf/policies/test_categorical_mlp_policy_with_model.py

```diff
@@ -1,54 +1,84 @@
 import pickle
+from unittest import mock
 
 import numpy as np
 import tensorflow as tf
 
 from garage.tf.envs import TfEnv
+from garage.tf.models import Model
 from garage.tf.policies import CategoricalMLPPolicyWithModel
 from tests.fixtures import TfGraphTestCase
 from tests.fixtures.envs.dummy import DummyDiscreteEnv
 
 
+class SimpleMLPModel(Model):
+    """Simple MLPModel for testing."""
+
+    def __init__(self, name, output_dim, *args, **kwargs):
+        super().__init__(name)
+        self.output_dim = output_dim
+
+    def _build(self, obs_input):
+        return tf.constant(
+            1., shape=(
+                1,
+                self.output_dim,
+            ))
+
+
 class TestCategoricalMLPPolicyWithModel(TfGraphTestCase):
     def setUp(self):
         super().setUp()
         self.env = TfEnv(DummyDiscreteEnv())
-        self.policy = CategoricalMLPPolicyWithModel(env_spec=self.env.spec)
         self.env.reset()
+        self.obs, _, _, _ = self.env.step(1)
+        with mock.patch(
+                'garage.tf.policies.categorical_mlp_policy_with_model.MLPModel',  # noqa: E501
+                new=SimpleMLPModel):
+            self.policy = CategoricalMLPPolicyWithModel(env_spec=self.env.spec)
 
     def test_get_action(self):
-        obs, _, _, _ = self.env.step(1)
-        action, _ = self.policy.get_action(obs)
+        action, _ = self.policy.get_action(self.obs)
         assert self.env.action_space.contains(action)
-        actions, _ = self.policy.get_actions([obs])
+
+        actions, _ = self.policy.get_actions([self.obs])
         for action in actions:
             assert self.env.action_space.contains(action)
 
-    def test_is_pickleable(self):
-        obs, _, _, _ = self.env.step(1)
-        action1, _ = self.policy.get_action(obs)
-        p = pickle.dumps(self.policy)
+    def test_dist_info(self):
+        policy_prob = self.policy.dist_info([self.obs])
```

Comment:
i don't understand this test. it seems that you're just verifying that `policy.dist_info` returns the same thing as `self.sess.run(...)` -- but that's a tautology.

you're testing the implementation.

the correct way to test this is to **know ahead-of-time** what the **actual numerical results** should be of `self.policy.dist_info` from `self.obs`. then you can make sure the numbers returned from `dist_info` are what you expet.

Reply:
Yes, I agree that is a tautology. I was wondering should I even test it.
However, since `self.policy.dist_info` is basically same as `self.sess.run(...)`, and `self.sess.run(...)` is actually the output of the underlying model. Numerical correctness is already tested in the tests of that model. Should I test it again here?

Reply:
yes. your test has no idea that there's a model inside the policy.